src/Storages/StorageURL.cpp

```diff
@@ -431,11 +446,27 @@ Chunk StorageURLSource::generate()
 
             progress(num_rows, chunk_size ? chunk_size : chunk.bytes());
             VirtualColumnUtils::addRequestedFileLikeStorageVirtualsToChunk(
-                chunk, requested_virtual_columns,
+                chunk,
+                requested_virtual_columns,
                 {
                     .path = curr_uri.getPath(),
                     .size = current_file_size,
-                }, getContext());
+                },
+                getContext());
+            if (need_headers_virtual_column)
+            {
+                if (!http_response_headers_initialized)
+                {
+                    http_response_headers = dynamic_cast<ReadWriteBufferFromHTTP *>(read_buf.get())->getResponseHeaders();
```

Comment:
Isn't it possible to have something other than `ReadWriteBufferFromHTTP` ? For example, when format is compressed and we have some wrapper on top?

Reply:
Why this is needed?

Reply:
It's just a guess that `read_buf` can be something other than `ReadWriteBufferFromHTTP`. Let's add an assert above and if tests passes it will mean that we can rely on that type